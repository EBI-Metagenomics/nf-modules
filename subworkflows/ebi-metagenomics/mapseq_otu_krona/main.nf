/* EBI-METAGENOMICS */
include { MAPSEQ             } from '../../../modules/ebi-metagenomics/mapseq/main'
include { MAPSEQ2BIOM        } from '../../../modules/ebi-metagenomics/mapseq2biom/main'
/* NF-CORE */
include { KRONA_KTIMPORTTEXT } from '../../../modules/nf-core/krona/ktimporttext/main'

workflow MAPSEQ_OTU_KRONA {

    take:
    ch_fasta    // channel: [ val(meta), [ bam ] ]
    ch_dbs      // channel: [ path(fasta), path(tax), path(otu), path(mscluster), val(label) ]

    main:

    ch_versions = Channel.empty()

    // db_fasta = ch_dbs[0]
    // db_tax = ch_dbs[1]
    // db_otu = ch_dbs[2]
    // db_mscluster = ch_dbs[3]
    // db_label = ch_dbs[4]

    ch_dbs.multiMap { fasta, tax, otu, mscluster, label ->
            mapseq_input: [fasta, tax, mscluster]
            mapseq_to_biom_input: [ otu, label ]
        }.set {
            input
        }


    MAPSEQ(
        ch_fasta,
        input.mapseq_input
    )
    ch_versions = ch_versions.mix(MAPSEQ.out.versions.first())

    MAPSEQ2BIOM(
        MAPSEQ.out.mseq,
        input.mapseq_to_biom_input
    )
    ch_versions = ch_versions.mix(MAPSEQ2BIOM.out.versions.first())

    KRONA_KTIMPORTTEXT(
        MAPSEQ2BIOM.out.krona_input
    )
    ch_versions = ch_versions.mix(KRONA_KTIMPORTTEXT.out.versions.first())

    emit:
    mseq                  = MAPSEQ.out.mseq                   // channel: [ val(meta), [ mseq ] ]
    krona_input           = MAPSEQ2BIOM.out.krona_input       // channel: [ val(meta), [ txt ] ]
    biom_out              = MAPSEQ2BIOM.out.biom_out          // channel: [ val(meta), [ tsv ] ]
    biom_notaxid_out      = MAPSEQ2BIOM.out.biom_notaxid_out  // channel: [ val(meta), [ tsv ] ]
    html                  = KRONA_KTIMPORTTEXT.out.html       // channel: [ val(meta), [ html ] ]
    versions              = ch_versions                       // channel: [ versions.yml ]
}
