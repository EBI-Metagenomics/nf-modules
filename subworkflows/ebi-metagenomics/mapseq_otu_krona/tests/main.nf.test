nextflow_workflow {

    name "Test Subworkflow MAPSEQ_OTU_KRONA"
    script "../main.nf"
    workflow "MAPSEQ_OTU_KRONA"
    config "./nextflow.config"


    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/mapseq_otu_krona"
    tag "mapseq"
    tag "mapseq2biom"
    tag "subworkflows/../../modules/nf-core/krona/ktimporttext"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall(
            "${launchDir}/library/",
            [
                "krona/ktimporttext",
            ]
        )
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")
    }

    test("test_mapseq_otu_krona") {

        when {
            workflow {
                """
                input[0] = [ [ id:'test', single_end:false ], // meta map
                        file("${moduleDir}/tests/subunit_silva_SSU.fasta", checkIfExists: true)
                        ]
                input[1] = Channel.of(
                            [
                            file("${moduleDir}/tests/silva_ssu/silva_ssu-20200130_SSU_trimmed.fasta", checkIfExists: true),
                            file("${moduleDir}/tests/silva_ssu/silva_ssu-20200130_SSU_filtered2.txt", checkIfExists: true),
                            file("${moduleDir}/tests/silva_ssu/silva_ssu-20200130_SSU_trimmed.otu", checkIfExists: true),
                            file("${moduleDir}/tests/silva_ssu/silva_ssu-20200130_SSU_trimmed.fasta.mscluster", checkIfExists: true),
                            "SILVA"
                            ]
                        )
                """
            }
        }
        then {
            assertAll(
                // These tests mainly check for headers rather than content because MAPseq is not deterministic.
                // This will change soon though so we can update these tests later.
                { assert workflow.success},
                { assert snapshot(path(workflow.out.mseq.get(0).get(1)).readLines()[0].contains("# mapseq v1.2.6 (Jan 20 2023)")).match("mapseq_output") },
                { assert snapshot(path(workflow.out.krona_input.get(0).get(1)).readLines()[0].contains("sk__Bacteria")).match("mapseq2biom_output_krona") },
                { assert snapshot(path(workflow.out.biom_out.get(0).get(1)).readLines()[0].contains("# Constructed from biom file")).match("mapseq2biom_output_biom") },
                { assert snapshot(path(workflow.out.biom_notaxid_out.get(0).get(1)).readLines()[0].contains("# Constructed from biom file")).match("mapseq2biom_output_biom_notaxid") },
                { assert snapshot(path(workflow.out.html.get(0).get(1)).readLines()[0].contains("DOCTYPE html PUBLIC")).match("krona_output") },
            )
        }
        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }
}
