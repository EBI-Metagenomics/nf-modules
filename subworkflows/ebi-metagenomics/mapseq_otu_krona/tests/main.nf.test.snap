{
    "mapseq2biom_output_krona": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-04-18T12:06:53.987371352"
    },
    "mapseq2biom_output_biom": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-04-18T12:06:54.014873914"
    },
    "mapseq_output": {
        "content": [
            false
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-21T15:02:21.978283709"
    },
    "mapseq2biom_output_biom_notaxid": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-04-18T12:06:54.025478915"
    },
    "krona_output": {
        "content": [
            true
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.3"
        },
        "timestamp": "2024-04-18T12:06:54.050582264"
    }
}